Domains: Add functionality to edit DNS records

leonardost opened the pull request:
### Changes proposed in this Pull Request

This PR updates the domain DNS management pages so it's possible to edit existing DNS records. When the "Edit" menu option is clicked in the record list page, you're taken to a "Edit DNS record" page where you can update the selected DNS record. The selected record's ID is passed via an URL query string parameter named `recordId`. This PR creates a new route (`/domains/manage/:domain/edit-dns-record/:site`) for this new DNS record edit page.

This is part of the domain pages redesign project described in pcYYhz-m2-p2. 

#### Screenshots:

Edit record menu option:

<img width="1085" alt="Screen Shot 2021-11-16 at 20 02 11" src="https://user-images.githubusercontent.com/5324818/142082337-19a5f7ca-22de-4a4a-a23b-1aa38432aa50.png">

Edit DNS record page:

<img width="1075" alt="Screen Shot 2021-11-16 at 20 03 00" src="https://user-images.githubusercontent.com/5324818/142082359-316b2617-07cb-4886-8f21-4529b5197479.png">

### Testing instructions

Build this branch locally or open the live Calypso link, go to "Upgrades > Domains" and select a registered domain. Go to its DNS management page and append `?flags=domains/dns-records-redesign` to your URL. Now you should be able to access the "Edit" option. 

Please try updating DNS records of different types and ensure they are updated correctly. Also try reloading while in the `edit-dns-record` page and ensure the record gets loaded correctly in that situation too. It would be nice to add a loading state in that scenario, while the record is being fetched from the backend, but I think we can do that in a later PR.

leonardost committed:
Go to "Add DNS record" page when clicking "Edit" menu of DNS record

leonardost committed:
Remove unnecessary code

leonardost committed:
Retrieve DNS record being edited in "edit DNS record" page

leonardost committed:
Load attributes of DNS record that is being edited

a8ci18n commented on client/my-sites/domains/domain-management/dns/add-dns-record.jsx:
```diff
@@ -97,24 +109,25 @@ class AddDnsRecprd extends Component {
 		);
+		const recordBeingEdited = this.getRecordBeingEdited();
+		const headerText = recordBeingEdited
+			? translate( 'Edit DNS record' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **29** times:
`translate( 'Edit DNS Record' )` *ES Score: 11*

leonardost committed:
Go back to DNS records list page after saving or editing a record

leonardost committed:
Dispatch `updateDns` action when updating a DNS record

a8ci18n commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -137,15 +163,34 @@ class DnsAddNew extends React.Component {
+				this.props.updateDns( selectedDomainName, [ normalizedData ], [ recordToEdit ] ).then(
+					() => {
+						page( domainManagementDns( selectedSiteSlug, selectedDomainName ) );
+						this.props.successNotice( translate( 'The DNS record has been updated.' ), {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **25** times:
`translate( 'The DNS record has been deleted.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-dns-records)

a8ci18n commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -137,15 +163,34 @@ class DnsAddNew extends React.Component {
+					},
+					( error ) => {
+						this.props.errorNotice(
+							error.message || translate( 'The DNS record has not been updated.' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'The DNS record has not been added.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-dns-records)

a8ci18n commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -208,6 +253,9 @@ class DnsAddNew extends React.Component {
 			record.types.includes( this.state.type )
 		);
+		const buttonLabel = recordToEdit
+			? translate( 'Update DNS record' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Update your DNS records' )` *ES Score: 7*

leonardost committed:
Do not clear fields when submitting form

leonardost committed:
Hide cancel button behind a feature flag

leonardost committed:
Fix form input warnings

Some form inputs in the MX and SRV records had an unnecessary `defaultValue`
attribute, which conflicts with the `value` property.

There were also some controlled/uncontrolled input warnings because forms for
all record types were actually always being rendered, just hidden with CSS -
the problem with that approach is that when the type was changed, properties
that existed for one record (e.g. "weight" for MX records) did not exist in
another (e.g. CNAME records). These attributes would become `undefined` in the
local state when the form was reset, causing the warnings.

leonardost committed:
Do not try to load DNS record when there's no record to load

leonardost committed:
Fix type in "AddDnsRecprd" component name

leonardost committed:
Get `selectedSite` from the global state

leonardost committed:
Refactor code to remove duplication

leonardost committed:
Simplify code

leonardost committed:
Remove code duplication

leonardost committed:
Set type in local state when loading record being edited

leonardost commented on client/my-sites/domains/domain-management/dns/add-dns-record.jsx:
```diff
@@ -25,16 +24,24 @@ import DnsAddNew from './dns-add-new';
 
 import './add-dns-record.scss';
 
-class AddDnsRecprd extends Component {
```
Typo

leonardost committed:
Rename query parameter "record" to "recordId"

a8ci18n commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -137,23 +166,39 @@ class DnsAddNew extends React.Component {
+
+			if ( recordToEdit ) {
+				this.props.updateDns( selectedDomainName, [ normalizedData ], [ recordToEdit ] ).then(
+					() => this.handleSuccess( translate( 'The DNS record has been updated.' ) ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **25** times:
`translate( 'The DNS record has been deleted.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-dns-records)

a8ci18n commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -137,23 +166,39 @@ class DnsAddNew extends React.Component {
+				this.props.updateDns( selectedDomainName, [ normalizedData ], [ recordToEdit ] ).then(
+					() => this.handleSuccess( translate( 'The DNS record has been updated.' ) ),
+					( error ) =>
+						this.handleError( error, translate( 'The DNS record has not been updated.' ) )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'The DNS record has not been disabled.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/edit-dns-records)

leonardost commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -180,34 +225,32 @@ class DnsAddNew extends React.Component {
+		return (
+			<selectedRecordType.component
+				selectedDomainName={ this.props.selectedDomainName }
+				show={ true }
```
I believe this property can now be removed from the record components, let's do that in a later PR.

leonardost committed:
Add new route `/domains/manage/:domain/edit-dns-record/:site` to edit records

delputnam commented on client/my-sites/domains/domain-management/dns/add-dns-record.jsx:
```diff
@@ -50,7 +57,9 @@ class AddDnsRecprd extends Component {
 			{
-				label: translate( 'Add a record', { comment: 'DNS record' } ),
+				label: recordBeingEdited
+					? translate( 'Edit record', { comment: 'DNS record' } )
```
Good catch! I forgot about this.

delputnam commented on client/my-sites/domains/domain-management/dns/dns-add-new.jsx:
```diff
@@ -119,16 +125,39 @@ class DnsAddNew extends React.Component {
+		const recordAttributes = Object.keys( selectedDnsRecordFields ).reduce( ( obj, field ) => {
+			obj[ field ] = recordToEdit[ field ];
+			return obj;
+		}, {} );
```
We'll need to check to see if this record is on the root domain here. Something like this maybe:
```suggestion
		const isRootDomainRecord = recordToEdit.domain === recordToEdit.name.replace( /\.$/, '' );
		const recordAttributes = Object.keys( selectedDnsRecordFields ).reduce( ( obj, field ) => {
			obj[ field ] = isRootDomainRecord && 'name' === field ? '' : recordToEdit[ field ];
			return obj;
		}, {} );
```

leonardost replied:
Thanks for the review @delputnam! I've created a method to process field values, as there were some other cases (`target` for SRV records, `data` in MX records) which also need some preprocessing.

delputnam reviewed (approved):
Nice work on this one!

In addition to the change to make sure we're setting the name field properly, I guess we'll want to somehow handle edits to the "protected" records until we get rid of them.  One way is to just add something like the following to `getActionsForDnsRecord`:
```
		if ( record.protected_field ) {
			return [];
		}
```
(as well as handle the case where we have an empty actions list by not showing the action menu icon for the record).

This would prevent editing for those records.

Or we could selectively handle this just for the wildcard CNAME record since that one shouldn't be deleted or modified.

The "protected" A record doesn't really include any data to be edited, but I suppose it's perhaps intuitive to let the user select it to be edited and then enter their own IP address.  If we decide to allow this case, we will still need to disable the "delete" action for this record I guess since it's not possible to actually delete the record without adding a new one.

Also, if we do allow editing of the protected A record, we'll also need to be sure to validate the data field since currently, that doesn't seem to happen.

My personal opinion is that we should just disable all actions on these protected records since the goal is to get rid of them eventually anyway.

leonardost committed:
Disable ellipsis menu for protected records

leonardost committed:
Process root records' name field and some SRV, MX and TXT fields too

We show users some DNS records values differently than they really are in raw
form. For example, root records have their name set as the domain name appended
with a period (`.`), and `target` (for SRV records) and `data` (for MX records)
contain hosts in FQDN form, i.e. also with a trailing period.

When we show these fields to users, we remove the trailing dot and, in the case
of the root record names, we simply do not show it.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6983643

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6983643).
